app/Http/Controllers/App/UsersController.php

```diff
@@ -36,58 +48,176 @@ public function add(Request $request)
 
     public function browse(Request $request)
     {
+        return self::json([], 501, ['message' => 'not yet implemented <3']);
         // TODO check privileges
-        $users = User::with('AdserverWallet')->whereNull('deleted_at')->get();
+        $users = User::with('AdserverWallet')->get();
 
         return self::json($users->toArrayCamelize());
     }
 
     public function delete(Request $request, $user_id)
     {
+        return self::json([], 501, ['message' => 'not yet implemented <3']);
         // TODO check privileges
         // TODO reset email
         // TODO process
-        $user = User::whereNull('deleted_at')->findOrFail($user_id);
-        $user->deleted_at = new \DateTime();
-        $user->save();
+        $user = User::findOrFail($user_id);
+        $user->delete();
 
-        return self::json(['message' => 'successfully deleted'], 200);
+        return self::json([], 204);
     }
 
-    public function edit(Request $request, $user_id)
+    public function edit(Request $request)
     {
-        // TODO check privileges
-        $user = User::whereNull('deleted_at')->findOrFail($user_id);
-        $this->validateRequest($request, 'user', User::$rules);
+        if (!Auth::check() && !$request->has('user.token')) {
+            return self::json([], 401, ['message' => 'Required authenticated access or token authentication']);
+        }
+
+        DB::beginTransaction();
+        if (Auth::check()) {
+            $user = Auth::user();
+            $token_authorization = false;
+        } else {
+            if (false === $token = Token::check($request->input('user.token'), null, 'password-recovery')) {
+                DB::rollBack();
+
+                return self::json([], 422, ['message' => 'Authentication token is invalid']);
+            }
+            $user = User::findOrFail($token['user_id']);
+            $token_authorization = true;
+        }
+
+        $this->validateRequestObject($request, 'user', User::$rules);
         $user->fill($request->input('user'));
+
+        if (!$request->has('user.password_new')) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "user.password_new" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)